cinder/volume/drivers/nexenta/ns5/iscsi.py

```diff
@@ -171,22 +166,37 @@ def delete_volume(self, volume):
 
         :param volume: volume reference
         """
-
-        url = ('storage/pools/%(pool)s/volumeGroups/%(group)s'
-               '/volumes/%(name)s') % {
-            'pool': self.storage_pool,
-            'group': self.volume_group,
-            'name': volume['name']
-        }
-        field = 'originalSnapshot'
-        origin = self.nef.get('{}?fields={}'.format(url, field)).get(field)
+        path = '%2F'.join([
+            self.storage_pool, self.volume_group, volume['name']])
+        url = 'storage/volumes/%s' % path
+        origin = self.nef.get(url).get('originalSnapshot')
         try:
+            url = 'storage/volumes/%s?snapshots=true' % path
             self.nef.delete(url)
         except exception.NexentaException as exc:
-            vol_path = self._get_volume_path(volume)
-            self.destroy_later_or_raise(exc, vol_path)
-            return
-        self.collect_zfs_garbage(origin)
+            if 'Failed to destroy snap' in exc.kwargs['message']['message']:
+                url = 'storage/snapshots?parent=%s' % path
+                snap_list = []
+                snap_map = {}
+                for snap in self.nef.get(url)['data']:
+                    snap_list.append(snap['path'])
+                for snap in snap_list:
+                    url = 'storage/snapshots/%s' % snap.replace('/', '%2F')
```

Comment:
Here and other places: I think Python must have something like quoteUrl to avoid manually replace all 'invalid' characters.

Reply:
'%2F'.join is a very common practice in python...

Reply:
It probably would make sense to change snap.replace('/', '%2F') to quote(snap, safe=''), but in case of join I don't see how we can utilize it. Instead of 
'%2F'.join([self.storage_pool, self.volume_group, volume])
we would get 
qoute('/'.join([self.storage_pool, self.volume_group, volume]))
which is just ugly